# **Aplicaciones Financieras de ML & AI**
## **Examen III:** *AI in Finance*

#### Nombre: Julio César Avila Torreblanca

- **Problema:**
    * Genere un modelo que dato un texto, nos regrese como predicción un salario estimado o un rango salarial estimado (use al menos 3 variables).
        - Use un modelo o variante de Bert.
    * Base de datos: consiste en Empleos, Descripciones de los empleos y Rango salarial.


- **Contenido del notebook**:
    1. Librerías y parámetros
    2. Lectura de datos
    3. EDA
    4. Procesamiento de datos
    5. Modelado
    6. Evaluación

# 1. Librerías y parámetros

In [39]:
# data
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split

from transformers import AutoModelForSeq2SeqLM, AutoTokenizer
import torch
from tqdm.auto import tqdm
from sentence_transformers import SentenceTransformer
from sentence_transformers.util import batch_to_device


# 2. Lectura de datos

In [40]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [43]:
df = pd.read_csv(
    '/content/drive/MyDrive/Academy/Diplomado_Finanazas&IA/01_Aplicaciones_Financieras_ML&AI/Tests/DataAnalyst.csv',
    #'data/DataAnalyst.csv',
    index_col=0,
    engine='python',
    encoding='utf-8',
)

df = df.reset_index(drop=True)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 262 entries, 0 to 261
Data columns (total 13 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Job Title          262 non-null    object 
 1   Salary Estimate    262 non-null    object 
 2   Job Description    262 non-null    object 
 3   Rating             262 non-null    float64
 4   Company Name       262 non-null    object 
 5   Location           262 non-null    object 
 6   Size               262 non-null    object 
 7   Type of ownership  262 non-null    object 
 8   Industry           262 non-null    object 
 9   Sector             262 non-null    object 
 10  Revenue            262 non-null    object 
 11  Competitors        262 non-null    object 
 12  Easy Apply         262 non-null    object 
dtypes: float64(1), object(12)
memory usage: 26.7+ KB


# 3. EDA
Tenemos un dataset con 262 registros sin valores nulos. En esta parte veremos columna a columna que tipo de valores tenemos, para definir las varaibles a considerar al modelo.

Columnas a evaluar:

- `Job Title`
- `Salary Estimate`
- `Job Description`
- `Rating`
- `Company Name`
- `Location`
- `Size`
- `Type of ownership`
- `Industry`
- `Sector`
- `Revenue`
- `Competitors`
- `Easy Apply`


In [57]:
df.head()

Job Title  \
0  Data Analyst, Center on Immigration and Justic...   
1                               Quality Data Analyst   
2  Senior Data Analyst, Insights & Analytics Team...   
3                                       Data Analyst   
4                             Reporting Data Analyst   

              Salary Estimate  \
0  $37K-$66K (Glassdoor est.)   
1  $37K-$66K (Glassdoor est.)   
2  $37K-$66K (Glassdoor est.)   
3  $37K-$66K (Glassdoor est.)   
4  $37K-$66K (Glassdoor est.)   

                                     Job Description  Rating  \
0  Are you eager to roll up your sleeves and harn...     3.2   
1  Overview\n\nProvides analytical and technical ...     3.8   
2  We’re looking for a Senior Data Analyst who ha...     3.4   
3  Requisition NumberRR-0001939\nRemote:Yes\nWe c...     4.1   
4  ABOUT FANDUEL GROUP\n\nFanDuel Group is a worl...     3.9   

                              Company Name      Location  \
0           Vera Institute of Justice\n3.2  New York, NY   
1  Visiting Nurse Service of New York\n3.8  New York, NY   
2                         Squarespace\n3.4  New York, NY   
3                            Celerity\n4.1  New York, NY   
4                             FanDuel\n3.9  New York, NY   

                     Size               Type of ownership  \
0    201 to 500 employees          Nonprofit Organization   
1        10000+ employees          Nonprofit Organization   
2  1001 to 5000 employees               Company - Private   
3    201 to 500 employees  Subsidiary or Business Segment   
4   501 to 1000 employees               Company - Private   

                           Industry                            Sector  \
0                 Social Assistance                        Non-Profit   
1  Health Care Services & Hospitals                       Health Care   
2                          Internet            Information Technology   
3                       IT Services            Information Technology   
4               Sports & Recreation  Arts, Entertainment & Recreation   

                      Revenue Competitors Easy Apply  
0  $100 to $500 million (USD)          -1       TRUE  
1      $2 to $5 billion (USD)          -1         -1  
2    Unknown / Non-Applicable     GoDaddy         -1  
3   $50 to $100 million (USD)          -1         -1  
4  $100 to $500 million (USD)  DraftKings       TRUE

In [58]:
df.describe(include='all')

Job Title             Salary Estimate  \
count            262                         262   
unique           169                          10   
top     Data Analyst  $43K-$76K (Glassdoor est.)   
freq              65                          31   
mean             NaN                         NaN   
std              NaN                         NaN   
min              NaN                         NaN   
25%              NaN                         NaN   
50%              NaN                         NaN   
75%              NaN                         NaN   
max              NaN                         NaN   

                                          Job Description      Rating  \
count                                                 262  262.000000   
unique                                                262         NaN   
top     Job Title :Senior SQL Data Analyst\n\nNo of Op...         NaN   
freq                                                    1         NaN   
mean                                                  NaN    3.103817   
std                                                   NaN    1.691081   
min                                                   NaN   -1.000000   
25%                                                   NaN    3.100000   
50%                                                   NaN    3.600000   
75%                                                   NaN    4.000000   
max                                                   NaN    5.000000   

           Company Name      Location               Size  Type of ownership  \
count               262           262                262                262   
unique              217            17                  9                 12   
top     Perficient\n3.6  New York, NY  1 to 50 employees  Company - Private   
freq                  5           214                 54                138   
mean                NaN           NaN                NaN                NaN   
std                 NaN           NaN                NaN                NaN   
min                 NaN           NaN                NaN                NaN   
25%                 NaN           NaN                NaN                NaN   
50%                 NaN           NaN                NaN                NaN   
75%                 NaN           NaN                NaN                NaN   
max                 NaN           NaN                NaN                NaN   

       Industry             Sector                   Revenue Competitors  \
count       262                262                       262         262   
unique       36                 18                        14          46   
top          -1  Business Services  Unknown / Non-Applicable          -1   
freq         47                 62                        76         207   
mean        NaN                NaN                       NaN         NaN   
std         NaN                NaN                       NaN         NaN   
min         NaN                NaN                       NaN         NaN   
25%         NaN                NaN                       NaN         NaN   
50%         NaN                NaN                       NaN         NaN   
75%         NaN                NaN                       NaN         NaN   
max         NaN                NaN                       NaN         NaN   

       Easy Apply  
count         262  
unique          2  
top            -1  
freq          246  
mean          NaN  
std           NaN  
min           NaN  
25%           NaN  
50%           NaN  
75%           NaN  
max           NaN

In [55]:
columns = [
  'Job Title',
  'Salary Estimate',
  'Job Description',
  'Rating',
  'Company Name',
  'Location',
  'Size',
  'Type of ownership',
  'Industry',
  'Sector',
  'Revenue',
  'Competitors',
  'Easy Apply',
]

df[columns].nunique()

Job Title            169
Salary Estimate       10
Job Description      262
Rating                28
Company Name         217
Location              17
Size                   9
Type of ownership     12
Industry              36
Sector                18
Revenue               14
Competitors           46
Easy Apply             2
dtype: int64

Para cada variable notemos lo siguiente:
- `Job Title`: título de la vacante, por los tipos de títulos existirán algunos que sean similares. Esto podría ser interesante para medir similitudes entre vacantes. (**Posible variable**)
- `Salary Estimate`: variable a predecir, podrían tomarse los diez intervalos como clases o hacer un procesamiento para obtener un valor continuo a predecir.
- `Job Description`: texto con la descripción del rol. (**Target**)
- `Rating`: valor continuo entre 0-5, en Glassdoor suele ser un rating de la vacante tomada por los usuarios. Este valor puede NO aportar al modelo.
- `Company Name`: nombre de la empresa, no aporta al modelo.
- `Location`: Lugar de la vacante, podría ser importante como variable. Pero debido a que son muy pocas vacantes y queremos generalizar, la descartaremos.
- `Size`: tamaño de la empresa, son pocas las categorías por lo que puede ser una buena variable a incluir en el modelo. (**Posible variable**)
- `Type of ownership`: tipo de empresa, dado que no son muchas categorías podría ser importante para el modelo. (**Posible variable**)
- `Industry`: industria,  dado que son demasiadas categorías lo descartaremos.
- `Sector`: sector al que pertenece, dado que no son muchas categorías podría ser importante para el modelo. (**Posible variable**)
- `Revenue`: intervalo de revenue de la empresa, podria ser importante para conocer las ganancias e impacto de la empresa. (**Posible variable**)
- `Competitors`: número de competidores, no aporta al modelo.
- `Easy Apply`: flag dado por la plataforma para la aplicación, no aporta al modelo.

De esta manera, consideraremos como vriables:
  - `Job Title`: txt a ser tratado con encoder.
  - `Job Description`: txt a ser tratado con encoder.
  - `Size`: txt a ser tratado con encoder.
  - `Type of ownership`: txt a ser tratado con encoder.
  - `Sector`: txt a ser tratado con encoder.
  - `Revenue`: txt a ser tratado con encoder.
  
Y como variable target: `Salary Estimate`.

# 4. Preprocesamiento de los datos
Haremos el siguiente procesamiento:
- Buscaremos predecir el promedio de los salario. Para ello haremos generaremos la target.
- Las variables categóricas, usaremos OHE donde solo consideraremos las categorías con mayor frecuencia.
- Para las variables `Job Title` y `Job Description` tomaremos dos encoders
  -

In [50]:
df['Salary Estimate'].value_counts(normalize=True)*100

Salary Estimate
$43K-$76K (Glassdoor est.)     11.832061
$37K-$66K (Glassdoor est.)     11.450382
$46K-$87K (Glassdoor est.)     11.450382
$51K-$88K (Glassdoor est.)     11.450382
$51K-$87K (Glassdoor est.)     11.450382
$59K-$85K (Glassdoor est.)     11.450382
$60K-$110K (Glassdoor est.)    11.450382
$41K-$78K (Glassdoor est.)     11.068702
$45K-$88K (Glassdoor est.)      4.198473
$73K-$127K (Glassdoor est.)     4.198473
Name: proportion, dtype: float64

In [14]:
df['Revenue'].value_counts()

Revenue
Unknown / Non-Applicable            76
$10+ billion (USD)                  26
$100 to $500 million (USD)          22
$1 to $5 million (USD)              20
$10 to $25 million (USD)            20
-1                                  16
$500 million to $1 billion (USD)    15
$25 to $50 million (USD)            15
$50 to $100 million (USD)           14
$2 to $5 billion (USD)              12
Less than $1 million (USD)          10
$1 to $2 billion (USD)               6
$5 to $10 billion (USD)              5
$5 to $10 million (USD)              5
Name: count, dtype: int64

In [15]:
df['Sector'].value_counts()

Sector
Business Services                     62
Information Technology                58
-1                                    47
Finance                               25
Health Care                           24
Media                                 11
Non-Profit                            10
Insurance                              5
Accounting & Legal                     5
Arts, Entertainment & Recreation       3
Education                              3
Consumer Services                      3
Restaurants, Bars & Food Services      1
Real Estate                            1
Retail                                 1
Government                             1
Biotech & Pharmaceuticals              1
Construction, Repair & Maintenance     1
Name: count, dtype: int64

In [16]:
df['Industry'].value_counts() # No

Industry
-1                                          47
IT Services                                 32
Staffing & Outsourcing                      24
Health Care Services & Hospitals            24
Investment Banking & Asset Management       17
Consulting                                  16
Internet                                    12
Advertising & Marketing                     11
Social Assistance                           10
Computer Hardware & Software                 8
Research & Development                       7
Enterprise Software & Network Solutions      6
Brokerage Services                           4
Insurance Carriers                           4
Accounting                                   4
Motion Picture Production & Distribution     4
Video Games                                  4
TV Broadcast & Cable Networks                3
Health, Beauty, & Fitness                    3
Colleges & Universities                      3
Gambling                                     2
Banks & Credit Unions                        2
Architectural & Engineering Services         2
Lending                                      1
Venture Capital & Private Equity             1
Casual Restaurants                           1
Building & Personnel Services                1
Sports & Recreation                          1
Food & Beverage Stores                       1
Legal                                        1
Real Estate                                  1
Federal Agencies                             1
Biotech & Pharmaceuticals                    1
Insurance Agencies & Brokerages              1
General Repair & Maintenance                 1
Security Services                            1
Name: count, dtype: int64

In [17]:
df['Type of ownership'].value_counts()

Type of ownership
Company - Private                 138
Company - Public                   54
Nonprofit Organization             28
-1                                 16
Subsidiary or Business Segment      9
Hospital                            6
Unknown                             4
Contract                            2
College / University                2
Government                          1
School / School District            1
Other Organization                  1
Name: count, dtype: int64

In [18]:
df['Size'].value_counts()

Size
1 to 50 employees          54
10000+ employees           45
1001 to 5000 employees     43
51 to 200 employees        39
201 to 500 employees       36
501 to 1000 employees      21
-1                         16
Unknown                     5
5001 to 10000 employees     3
Name: count, dtype: int64

## Getting train/test

In [19]:
df.loc[:,['salary_interval']] = df['Salary Estimate'].apply(lambda x: x.split(' ')[0])
df.loc[:,['salary_interval']]

salary_interval
0         $37K-$66K
1         $37K-$66K
2         $37K-$66K
3         $37K-$66K
4         $37K-$66K
..              ...
257      $73K-$127K
258      $73K-$127K
259      $73K-$127K
260      $73K-$127K
261      $73K-$127K

[262 rows x 1 columns]

In [20]:
df.loc[:,['salary_min']] = df['salary_interval'].apply(lambda x: x.split('-')[0])
df.loc[:,['salary_min']] = df['salary_min'].apply(lambda x: x.replace('$',''))
df.loc[:,['salary_min']] = df['salary_min'].apply(lambda x: x.replace('K',''))
df.loc[:,['salary_min']] = df.loc[:,['salary_min']].astype(int)*1000
df.loc[:,['salary_min']]

salary_min
0        37000
1        37000
2        37000
3        37000
4        37000
..         ...
257      73000
258      73000
259      73000
260      73000
261      73000

[262 rows x 1 columns]

In [21]:
df.loc[:,['salary_max']] = df['salary_interval'].apply(lambda x: x.split('-')[1])
df.loc[:,['salary_max']] = df['salary_max'].apply(lambda x: x.replace('$',''))
df.loc[:,['salary_max']] = df['salary_max'].apply(lambda x: x.replace('K',''))
df.loc[:,['salary_max']] = df.loc[:,['salary_max']].astype(int)*1000
df.loc[:,['salary_max']]

salary_max
0        66000
1        66000
2        66000
3        66000
4        66000
..         ...
257     127000
258     127000
259     127000
260     127000
261     127000

[262 rows x 1 columns]

In [22]:
cols = [
    'Job Title',
    'Job Description',
    'Size',
    'Type of ownership',
    'Industry',
    'Sector',
]

X_train = df.loc[:,cols].copy()
y_train = df.loc[:,['salary_min']].copy()

In [23]:
X_train

Job Title  \
0    Data Analyst, Center on Immigration and Justic...   
1                                 Quality Data Analyst   
2    Senior Data Analyst, Insights & Analytics Team...   
3                                         Data Analyst   
4                               Reporting Data Analyst   
..                                                 ...   
257                                  Data Analyst - QC   
258                   People Operations & Data Analyst   
259                        Lead Data Analyst (Product)   
260                                 Data Analyst - III   
261                                   SQL Data Analyst   

                                       Job Description  \
0    Are you eager to roll up your sleeves and harn...   
1    Overview\n\nProvides analytical and technical ...   
2    We’re looking for a Senior Data Analyst who ha...   
3    Requisition NumberRR-0001939\nRemote:Yes\nWe c...   
4    ABOUT FANDUEL GROUP\n\nFanDuel Group is a worl...   
..                                                 ...   
257  Nesco Resource is seeking a Data Analyst for a...   
258  JOB DESCRIPTION:\n\nMuseum of Ice Cream is see...   
259  A BIT ABOUT OUR DATA & ANALYTICS TEAM\n\nThe K...   
260  Direct Client Requirement\nPosition: Data Anal...   
261  Job Title :Senior SQL Data Analyst\n\nNo of Op...   

                       Size               Type of ownership  \
0      201 to 500 employees          Nonprofit Organization   
1          10000+ employees          Nonprofit Organization   
2    1001 to 5000 employees               Company - Private   
3      201 to 500 employees  Subsidiary or Business Segment   
4     501 to 1000 employees               Company - Private   
..                      ...                             ...   
257  1001 to 5000 employees               Company - Private   
258    201 to 500 employees               Company - Private   
259       1 to 50 employees               Company - Private   
260       1 to 50 employees               Company - Private   
261     51 to 200 employees               Company - Private   

                             Industry                            Sector  
0                   Social Assistance                        Non-Profit  
1    Health Care Services & Hospitals                       Health Care  
2                            Internet            Information Technology  
3                         IT Services            Information Technology  
4                 Sports & Recreation  Arts, Entertainment & Recreation  
..                                ...                               ...  
257            Staffing & Outsourcing                 Business Services  
258                                -1                                -1  
259           Advertising & Marketing                 Business Services  
260           Advertising & Marketing                 Business Services  
261                        Consulting                 Business Services  

[262 rows x 6 columns]

In [24]:
from transformers import AutoModelForSeq2SeqLM, AutoTokenizer

tokenizer = AutoTokenizer.from_pretrained("avisena/bart-base-job-info-summarizer")
model = AutoModelForSeq2SeqLM.from_pretrained("avisena/bart-base-job-info-summarizer")

/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


tokenizer_config.json:   0%|          | 0.00/1.43k [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/798k [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/2.11M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/957 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/1.77k [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/558M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/292 [00:00<?, ?B/s]

In [25]:
input_text = X_train.iloc[0]['Job Description']
input_text

"Are you eager to roll up your sleeves and harness data to drive policy change? Do you enjoy sifting through complex datasets to illuminate trends and insights? Do you see yourself working for a values-driven organization with a vision to tackle the most pressing injustices of our day?\n\nWe are looking to hire a bright, hard-working, and creative individual with strong data management skills and a demonstrated commitment to immigrant's rights. The Data Analyst will assist with analysis and reporting needs for Veras Center on Immigration and Justice (CIJ), working across its current projects and future Vera initiatives.\n\nWho we are:\n\nFounded in 1961, The Vera Institute is an independent, non-partisan, nonprofit organization that combines expertise in research, technical assistance, and demonstration projects to assist leaders in government and civil society examine justice policy and practice, and improve the systems people rely on for justice and safety.\nWe study problems that im

In [26]:
inputs = tokenizer.encode(input_text, return_tensors="pt", max_length=1024, truncation='do_not_truncate')


In [29]:
inputs.shape

torch.Size([1, 1350])

In [30]:
"""summary_ids = model.generate(
    inputs,
    max_length=200,  # Maximum length of the summary
    min_length=30,   # Minimum length of the summary
    length_penalty=0.98,  # Penalty for longer sequences
    num_beams=6,     # Number of beams for beam search
    top_p=3.7,
    early_stopping=True,
    temperature=1.4,
    do_sample=True
)

summary = tokenizer.decode(summary_ids[0], skip_special_tokens=True, max_length=512, truncation='do_not_truncate')

print(f"Generated Summary: {summary}")"""

'summary_ids = model.generate(\n    inputs, \n    max_length=200,  # Maximum length of the summary\n    min_length=30,   # Minimum length of the summary\n    length_penalty=0.98,  # Penalty for longer sequences\n    num_beams=6,     # Number of beams for beam search\n    top_p=3.7,\n    early_stopping=True,\n    temperature=1.4,\n    do_sample=True\n)\n\nsummary = tokenizer.decode(summary_ids[0], skip_special_tokens=True, max_length=512, truncation=\'do_not_truncate\')\n\nprint(f"Generated Summary: {summary}")'

In [31]:
import torch
import numpy as np
from tqdm.auto import tqdm
from sentence_transformers import SentenceTransformer
from sentence_transformers.util import batch_to_device, cos_sim

# Load the model
model = SentenceTransformer("TechWolf/JobBERT-v2")

def encode_batch(jobbert_model, texts):
    features = jobbert_model.tokenize(texts)
    features = batch_to_device(features, jobbert_model.device)
    features["text_keys"] = ["anchor"]
    with torch.no_grad():
        out_features = jobbert_model.forward(features)
    return out_features["sentence_embedding"].cpu().numpy()

def encode(jobbert_model, texts, batch_size: int = 8):
    # Sort texts by length and keep track of original indices
    sorted_indices = np.argsort([len(text) for text in texts])
    sorted_texts = [texts[i] for i in sorted_indices]

    embeddings = []

    # Encode in batches
    for i in tqdm(range(0, len(sorted_texts), batch_size)):
        batch = sorted_texts[i:i+batch_size]
        embeddings.append(encode_batch(jobbert_model, batch))

    # Concatenate embeddings and reorder to original indices
    sorted_embeddings = np.concatenate(embeddings)
    original_order = np.argsort(sorted_indices)
    return sorted_embeddings[original_order]



modules.json:   0%|          | 0.00/339 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/201 [00:00<?, ?B/s]

README.md:   0%|          | 0.00/16.3k [00:00<?, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/618 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/438M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/1.58k [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/711k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/964 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/296 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/115 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/333 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/3.15M [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/3.15M [00:00<?, ?B/s]

NameError: name 'job_titles' is not defined

In [32]:
# Example usage
job_titles = [X_train.iloc[0]['Job Title']]

# Get embeddings
embeddings = encode(model, job_titles)

  0%|          | 0/1 [00:00<?, ?it/s]

In [35]:
embeddings.shape

(1, 1024)

In [37]:
embeddings

array([[-0.01685412,  0.05697987, -0.00716322, ...,  0.10164313,
        -0.04248427, -0.03711392]], dtype=float32)

In [36]:
inputs.shape

torch.Size([1, 1350])

In [38]:
inputs

tensor([[    0, 13755,    47,  ...,  9529, 16271,     2]])